images/airflow/2.10.1/python/mwaa/celery/task_monitor.py

```diff
@@ -474,35 +559,60 @@ def process_next_signal(self):
                 "after it has been closed."
             )
             return
-        signal_filepath, signal_data = _get_next_unprocessed_signal()
+        signal_filepath, signal_data = self._get_next_unprocessed_signal()
         if not signal_data:
             logger.info("No new signal found.")
             return
-        signal_id = signal_data["executionId"]
-        signal_type = signal_data["signalType"]
-        signal_timestamp = signal_data["createdAt"]
+        signal_id = signal_data.executionId
+        signal_type = signal_data.signalType
+        signal_timestamp = signal_data.createdAt
         logger.info(f"Processing signal {signal_id} of type {signal_type} created at {signal_timestamp}")
-        if signal_type == "activation":
-            self.waiting_for_activation = False
-        elif signal_type == "kill":
-            self.marked_for_kill = True
-        elif signal_type == "termination":
-            if (self.last_termination_or_resume_signal_timestamp is None or
-                    self.last_termination_or_resume_signal_timestamp < signal_timestamp):
-                self.marked_for_termination = True
-                self.last_termination_or_resume_signal_timestamp = signal_timestamp
-                self.last_termination_processing_time = datetime.now(tz=tz.tzutc())
-        elif signal_type == "resume":
-            if (self.last_termination_or_resume_signal_timestamp is None or
-                    self.last_termination_or_resume_signal_timestamp < signal_timestamp):
-                self.marked_for_termination = False
-                self.last_termination_or_resume_signal_timestamp = signal_timestamp
-                self.last_termination_processing_time = None
-        else:
-            logger.warning(f"Unknown signal type {signal_type}, ignoring.")
+        self._process_signal(signal_type, signal_timestamp)
+        self._marked_signal_as_processed(signal_filepath, signal_data)
+
+    def _process_signal(self, signal_type, signal_timestamp):
+        """
+        This method is used to process a signal. It will update the state of work consumption accordingly.
+        :param signal_type: Type of the signal received.
+        :param signal_timestamp: Timestamp at which the signal was generated.
+        :return:
+        """
+        signal_ignored = 0
+        match signal_type:
+            case SignalType.ACTIVATION:
+                # Skipping checking if activation is already present because activation signals are by
+                # design sent in redundant fashion to improve success of a newly created worker from getting
+                # activated by the worker enabler lambda.
+                self.waiting_for_activation = False
+            case SignalType.KILL:
+                if self.marked_for_kill:
+                    logger.warning("Received kill signal but already marked for kill. Ignoring.")
+                    signal_ignored = 1
```

Comment:
Is the signal ignored in this case? You're re-applying on the line below.

Reply:
My intention behind signal_ignored was to catch the unknown cases where we have redundant signals coming in. Here, we have a kill signal coming in for a worker which has already been marked for killing from a previous kill signal. Hence, the new signal is redundant and we emit signal_ignored = 1.